### ISSUE TITLE
Common imports.

### ISSUE THREAD

**Issue by lukesneeringer** [googleapis/api-linter#84]
We need to be able to resolve imports for common protos (e.g. `google/api/annotations.proto`) in order to write many rules.

I assert that any protos defined in [this repository](https://github.com/googleapis/api-common-protos) should be able to be imported in test templates.

**Comment by mingzhi** [comment]
Agree. But we need to be find a way to have a local (and synced) copy of the mentioned repository (we have one in Google3).

**Comment by lukesneeringer** [comment]
That repository downsyncs from google3. :) It copies all the directories listed in [AIP-213](https://aip.dev/213).

**Comment by mingzhi** [comment]
Rob, can you please take a look at it? Thank you!

**Comment by lukesneeringer** [comment]
@robcapo I could believe that this might be tricky. If you want, I can ask some of the Golang heavy hitters on my team to help, since they have experience writing golang / protoc stuff.

**Comment by robcapo** [comment]
I have some thoughts (and open questions) on the different ways we can do this, and I'm actually writing a test in google3 that needs it now, so it's become priority 1 for me. I think your idea about a submodule externally makes sense, and we can just use the corresponding google3 paths internally.

The mechanics of passing multiple files into protoc is also straightforward. We can either compile the entire repository into a descriptor set and pass that in, or we can pass the base directories in via proto_path.

If you want to reach out to the Golang team and get some advice, that'd be great. In the meantime I'll try to throw together some prototypes and see if we can get something working (since I'm blocked by it anyway). Thanks!

**Comment by lukesneeringer** [comment]
Adding @noahdietz and @pongad for vis -- Noah and Michael, if you have advice for us on this, we would really appreciate it.

**Comment by noahdietz** [comment]
Let me reiterate what I think is being discussed and y'all tell me if I'm off:

* `api-common-protos` is a dependency for many of the rules that will be written (e.g. validating a `google.api.resource` annotation).
* It would be nice if the linter could provide the `api-common-protos` (for both binary & Go package usage) so that a user doesn't have to clone the repo & provide them (e.g. via `--proto_path`). 

The questions are:
* how do we package/ship `api-common-protos` with a linter binary?
* how do we expose `api-common-protos` to the linter (pkg & binary protoc exec)?

**Comment by lukesneeringer** [comment]
> It would be nice if the linter could provide the `api-common-protos` (for both binary & Go package usage) so that a user doesn't have to clone the repo & provide them (e.g. via `--proto_path`).

Close. The bigger issue is that the _tests for the linter rules_ often need to be able to import this stuff to work. For example, let's say I want to write a test for [AIP-131](https://aip.dev/131) that tests that it actually uses the GET HTTP method.

The proper proto looks like...

```proto
syntax = "proto3";

import "google/api/annotations.proto";

service DummyAip131 {
  rpc GetFoo(GetFooRequest) returns (Foo) {
    option (google.api.http) = {
      get: "/v1/{name=bars/*/foos/*}"
    };
  }
}

// ...and obviously the messages referenced above
```

Right now, I can not write this in [a test like this one](https://github.com/googleapis/api-linter/blob/aip-131/rules/0131_test.go#L23) because the testing framework actually compiles the proto and sends a descriptor, and it has no way to resolve that import.

That said, this might be a distinction without a difference -- the fundamental problem is the same -- the linter needs access to our standard library, and your questions are otherwise spot on.

**Comment by noahdietz** [comment]
Ok thanks for the clarification Luke. I looked at what the `testutil` pkg does to convert a string template to a `FileDescriptorProto` and I've got some ideas. However, I think there should be a separation of concerns between how a *test* constructs a suitable `FileDescriptorProto` (w/imported protos) and how the *binary* (or CLI I guess) supplies the `api-common-protos`.

For the "shipping a binary" case, we can't compile a Go binary with static files in it (w/o extra tooling e.g. https://github.com/gobuffalo/packr) and then reference them on the host machine file system from an exec'd command (unless we packed the files via packr and wrote them to a tmp dir, including *that* tmp dir via `--proto_path`). I guess you could do all of those things with a descriptor packr'd into the binary and written to a temp dir on the user host.

As for testing, I don't mean to shake things up/critique out of nowhere, but I think writing a string to a tmp file and running protoc against it to then load the descriptor from file is a lot of moving parts for test data setup. I've got 2 ideas:

1. Keep a `testdata` folder with `.proto` files and their compiled `.pb.go` files checked in. Compile them as part of code changes/releases with `api-common-protos` on your machine. From tests either:
  **a.** load the proto file directly (via [proto.FileDescriptor](https://godoc.org/github.com/golang/protobuf/proto#FileDescriptor)) and unmarshal it into a FileDescriptorProto. I did this in tests for a validator of my own (with the help of jhump/protoreflect for loading) [here](https://github.com/googleapis/gapic-config-validator/blob/83e4f8f68adc0559a4708d94a0a43e21e103bb38/internal/validator/validator_test.go#L310).
  **b.** using the `Descriptor() ([]byte, error)` Go API (from the `.pb.go` files), unmarshal that into a FileDescriptorProto

2. Build the FileDescriptorProtos on the fly using [github.com/jhump/protoreflect/desc/builder](https://github.com/jhump/protoreflect/desc/builder). No checked in `.proto`/`.pb.go` files. If you set an option on a message using code, its descriptor is included. I did this in a plugin behavior validation tool I wrote. [Load `api-common-protos` descriptors](https://github.com/googleapis/gapic-config-validator/blob/83e4f8f68adc0559a4708d94a0a43e21e103bb38/cmd/gapic-error-conformance/main.go#L344-L345). [Build full CodeGeneratorRequests](https://github.com/googleapis/gapic-config-validator/blob/83e4f8f68adc0559a4708d94a0a43e21e103bb38/cmd/gapic-error-conformance/main.go#L144-L146).

If we are sticking with running `protoc` inside of `testutil` I think @robcapo's suggestion of checking in a descriptor set to pass in is reasonable.

Happy to talk about this more offline as well if you think the above suggestions are interesting.

**Comment by alin04** [comment]
Drive-by (I haven't read all the comments), so apologies if I am completely off-base. Would installing a dependency in the vendor directory help at all?

**Comment by robcapo** [comment]
> Ok thanks for the clarification Luke. I looked at what the `testutil` pkg does to convert a string template to a `FileDescriptorProto` and I've got some ideas. However, I think there should be a separation of concerns between how a _test_ constructs a suitable `FileDescriptorProto` (w/imported protos) and how the _binary_ (or CLI I guess) supplies the `api-common-protos` (to the underlying protoc invocation).

I agree, I think. Basically, we have the testutil package doing similar things to what the cli is doing. Both are executing protoc under the hood, but they are not tied to each other in any way. That said, if we, for example, checked in a descriptor set containing the `api-common-protos`, I don't think there's anything wrong with having both the CLI and the testutil package use the same descriptor set.
> For the "shipping a binary" case, we can't compile a Go binary with static files in it (w/o extra tooling e.g. https://github.com/gobuffalo/packr) and then reference them on the host machine file system from an exec'd command (unless we packed the files via packr and wrote them to a tmp dir, including _that_ tmp dir via `--proto_path`). I guess you could do all of those things with a descriptor packr'd into the binary and written to a temp dir on the user host.

I'm not super concerned with this at the moment. Particularly inside of google3, it's not a concern. We actually aren't even building the protos (nor using the cli), but receiving blaze build events (through Tricorder) with the descriptor sets of the protos we're linting. Therefore, we don't even need the `api-common-protos`, because they will already be available inside the descriptor sets that we receive from blaze. Outside of google3, I'm not familiar enough with the usage to have an opinion, but I'm happy to discuss this more. @lukesneeringer and @jgeewax are definitely more familiar with the use-cases outside of google3.
> As for testing, I don't mean to shake things up/critique out of nowhere, but I think writing a string to a tmp file and running protoc against it to then load the descriptor from file is a lot of moving parts for test data setup. I've got 2 ideas:

Yeah, it's not my favorite either. The main problem is that we need source code info for the linter `Rule`s to post locations, but we need the `protoreflect.Descriptor` representation in order to effectively traverse and inspect proto defs.
> 
> 1. Keep a `testdata` folder with `.proto` files and their compiled `.pb.go` files checked in. Compile them as part of code changes/releases with `api-common-protos` on your machine. From tests either:

My concern with this approach (we started with this) is that it becomes very cumbersome to write tests. If you're writing a simple rule, it's nice to be able to use [text/template](https://golang.org/pkg/text/template/) to write a few test cases without the burden of not only adding a file for each case, but also compiling that file and keeping it up to date. There's a trade-off here between the complexity of writing a test and the complexity of the underlying test infrastructure, and I think we decided that the former is more important than the latter.
>    **a.** load the proto file directly (via [proto.FileDescriptor](https://godoc.org/github.com/golang/protobuf/proto#FileDescriptor)) and unmarshal it into a FileDescriptorProto. I did this in tests for a validator of my own (with the help of jhump/protoreflect for loading) [here](https://github.com/googleapis/gapic-config-validator/blob/83e4f8f68adc0559a4708d94a0a43e21e103bb38/internal/validator/validator_test.go#L310).
>    **b.** using the `Descriptor() ([]byte, error)` Go API (from the `.pb.go` files), unmarshal that into a FileDescriptorProto
> 2. Build the FileDescriptorProtos on the fly using [github.com/jhump/protoreflect/desc/builder](https://github.com/jhump/protoreflect/desc/builder). No checked in `.proto`/`.pb.go` files. If you set an option on a message using code, its descriptor is included. I did this in a plugin behavior validation tool I wrote. [Load `api-common-protos` descriptors](https://github.com/googleapis/gapic-config-validator/blob/83e4f8f68adc0559a4708d94a0a43e21e103bb38/cmd/gapic-error-conformance/main.go#L344-L345). [Build full CodeGeneratorRequests](https://github.com/googleapis/gapic-config-validator/blob/83e4f8f68adc0559a4708d94a0a43e21e103bb38/cmd/gapic-error-conformance/main.go#L144-L146).

I don't think I fully understand what's happening here, but this might be an interesting solution. I think, again, focusing on keeping tests lightweight, the best solution is the simplest implementation that allows us to start with proto source as a Go string, and get a FileDescriptorProto with source code info out. Note that we won't need source code info for the `api-common-protos` descriptors though, because those aren't the ones we'll be linting.

**Comment by noahdietz** [comment]
> I'm not super concerned with this at the moment. Particularly inside of google3, it's not a concern.

Fair, it's a separate issue that can be discussed elsewhere. Putting aside the "expose common protos to the CLI" case and focusing on tests.

> The main problem is that we need source code info for the linter Rules to post locations, but we need the protoreflect.Descriptor representation in order to effectively traverse and inspect proto defs

Ah I'm not sure that the library I've been suggesting supplies source code **location** information. That would preclude pretty much everything I've suggested I think.

> My concern with this approach (we started with this) is that it becomes very cumbersome to write tests.

Fair, point taken. Discount suggestion 1.

> I don't think I fully understand what's happening here, but this might be an interesting solution

Sorry if the code/explanation wasn't clear, I didn't want to be too verbose in an issue comment. Focusing on suggestion 2. jhump/protoreflect/desc/builder provides "proto builders" so that you can build new descriptors in code, and assemble them into a FileDescriptorProto at the end. Instead of writing a `.proto` template string to a file & compiling a descriptor w/protoc, it can be assembled entirely in code with the builders.

This would be more code in the tests, and *could* be harder to understand at first glance, I admit that. However, it is more flexible and eliminates a test dependency on `protoc` (and what seems inevitable, a dep on `api-common-protos` as a submodule or checked-in descriptor set). I also think it scales a bit better for tests spanning multiple files (e.g. linting a `google.api.resource_reference` or `google.longrunning.operation_info.response_type`), but I haven't thought deeply about this and I'm happy to be disproved. 

Again, I don't want to shake anything up, just sharing different approaches. :smiley: 

I think checking in an `api-common-protos` descriptor set for **test** data setup makes sense. But it might easier to maintain a submodule rather than having to compile & check-in a descriptor set.

> Would installing a dependency in the vendor directory help at all?

@alin04 I don't think it's a Go dependency but rather a proto one. The `api-common-protos` should be available for the rules/testutil code that runs `protoc` to compile test data. This way, it's easier to write tests for rules that lint annotations defined in `api-common-protos`. If my understanding of the suggestion is off, let me know :sweat_smile:

**Comment by lukesneeringer** [comment]
> I think, again, focusing on keeping tests lightweight, the best solution is the simplest implementation that allows us to start with proto source as a Go string, and get a FileDescriptorProto with source code info out.

+1

Going back to the dependency thing, how do we want to handle this? Submodule? Vendoring? Something else? Agree that it is a non-issue in internal source control.

**Comment by pongad** [comment]
I have very little context on this, so apologies if I'm missing something obvious.

I agree with trade off discussions in this thread. Of all proposals discussed, I think submodule is probably the least annoying solution.

One suggestion that might be completely out the left field: is it possible to lint without compiling the proto file at all? Looking at the protobuf syntax documentation, it seems possible to write relatively sane-looking yacc grammar for it. Looking through AIPs, many of them can probably be checked in this way and we can forget about `protoc` altogether.

I'm not sure how well this plays with google3 tools though.

**Comment by mingzhi** [comment]
> I have very little context on this, so apologies if I'm missing something obvious.
> 
> I agree with trade off discussions in this thread. Of all proposals discussed, I think submodule is probably the least annoying solution.
> 
> One suggestion that might be completely out the left field: is it possible to lint without compiling the proto file at all? Looking at the protobuf syntax documentation, it seems possible to write relatively sane-looking yacc grammar for it. Looking through AIPs, many of them can probably be checked in this way and we can forget about `protoc` altogether.

Parsers based on yacc grammar are totally doable and not that hard as Protobuf has [EBNF spec](https://developers.google.com/protocol-buffers/docs/reference/proto3-spec). But stitching dependencies together would be a nightmare.
> 
> I'm not sure how well this plays with google3 tools though.

Google3 is a beast on the cloud (we spent several weeks on figuring out how to include imports in protoc in Google3, as all files are on the cloud).

**Comment by pongad** [comment]
Do you actually need to stitch together the dependencies though? I can imagine you could do something like looking at the annotations for a method and make sure it has `google.api.http` that has a `post` field, etc.

If the dependencies aren't on the disk, then we let the linting pass and let the build fail. To me, this kind of thing seems reasonable, but I seem to sometimes have strange expectations :P

**Comment by lukesneeringer** [comment]
I think the question is at least interesting enough to be asked, but I expect we want to use descriptors for testing. String parsing (1) seems like not a very real test (although maybe for unit that is fine) and (2) seems likely to need to be constantly updated.

**Comment by pongad** [comment]
(1) I share the feeling, though I struggle to explain it: what's the "virtual" part of it, exactly? Is it that lint might succeed even though compilation fails? 

(2) Why do you think it needs to be updated more regularly than using descriptors? It is new proto file syntaxes?

**Comment by lukesneeringer** [comment]
> EDIT: An unstated motivation. If we can deal with linting completely syntactically, it becomes much easier to turn the linter into some kind of language server we can integrate into IDEs. Then the language server doesn't need to understand the user's build system to do its job. This probably won't make or break the deal for you but it sounds like a good possibility to entertain.

This is a non-trivially interesting point. As in, I really wish I had thought of it months ago.

That said, I question whether it can be easily done at this point (but I do not know as much as Rob or Mingzhi).

**Comment by pongad** [comment]
Totally understand. I don't think you need to backtrack to support this. If we need it for something in the future, I'll come knocking :D

**Comment by mingzhi** [comment]
About syntax parsers, I did some research a while ago and even wrote a simple parser for it, but decided not to go that direction due to: 
1. Google3 is currently our biggest customer, it is not trivial to solve dependencies using those parsers; (actually it is super hard, we knew that a colleague spent 14 months of 20% time to build a service to solve some dependency issues using `protoc`).
2. This API linter does not just check protobuf syntax, but also review the whole picture of a service definition -- however, we cannot assume that an API service is defined in a single file without any other imported files, e.g., a `UserInfoService` in the `user_info_service.proto` may import a `UserInfo` message defined in the `user_info.proto`; and without dependencies, it is impossible to check if the `UserInfo` message has some required fields.

In short, if dependency is to be resolved, I don't see any other parser can beat `protoc`. 

However, I was also trying to decouple parsers from the linter, so that different parsers can be used in the future. And I love to see protobuf compiler written in Go (I like jhump's parser, but unfortunately we haven't found a way to use it in Google3).

Thanks!!!

**Comment by mingzhi** [comment]
Back to common (standard) protobuf files, I would like to point out that the API linter doesn't check those files and it doesn't make sense for any rule to do so, e.g., we don't care if the message `Date` in `google/type/date.proto` is good or not in style (I think we can assume that it is so). And therefore, those common protobuf files are only for `protoc` to complete compilation, and we don't need source code information or dependencies for them.

So I agree with Noah that pre-compiled descriptor in the Go packages is a valid solution for including those common protobuf files into the linter/test binary. E.g.,
```
import _ "google.golang.org/genproto/googleapis/type/date" // for "google/type/date.proto"
```
and then get the gzipped `FileDescriptorProto` for `google/type/date.proto`
```
gzippedBytes := proto.FileDescriptor("google/type/date.proto")
```

**Comment by pongad** [comment]
@mingzhi Thank you for the context! I agree that with the given constraints, the syntax-based approach makes less sense.
